programs/client/Suggest.cpp

```diff
@@ -90,45 +91,69 @@ void Suggest::loadImpl(Connection & connection, const ConnectionTimeouts & timeo
     /// NOTE: Once you will update the completion list,
     /// do not forget to update 01676_clickhouse_client_autocomplete.sh
 
-    std::stringstream query;        // STYLE_CHECK_ALLOW_STD_STRING_STREAM
-    query << "SELECT DISTINCT arrayJoin(extractAll(name, '[\\\\w_]{2,}')) AS res FROM ("
-        "SELECT name FROM system.functions"
-        " UNION ALL "
-        "SELECT name FROM system.table_engines"
-        " UNION ALL "
-        "SELECT name FROM system.formats"
-        " UNION ALL "
-        "SELECT name FROM system.table_functions"
-        " UNION ALL "
-        "SELECT name FROM system.data_type_families"
-        " UNION ALL "
-        "SELECT name FROM system.merge_tree_settings"
-        " UNION ALL "
-        "SELECT name FROM system.settings"
-        " UNION ALL "
-        "SELECT cluster FROM system.clusters"
-        " UNION ALL "
-        "SELECT name FROM system.errors"
-        " UNION ALL "
-        "SELECT event FROM system.events"
-        " UNION ALL "
-        "SELECT metric FROM system.asynchronous_metrics"
-        " UNION ALL "
-        "SELECT metric FROM system.metrics"
-        " UNION ALL "
-        "SELECT macro FROM system.macros"
-        " UNION ALL "
-        "SELECT policy_name FROM system.storage_policies"
-        " UNION ALL "
-        "SELECT concat(func.name, comb.name) FROM system.functions AS func CROSS JOIN system.aggregate_function_combinators AS comb WHERE is_aggregate";
+     Settings settings;
+    /// To show all rows from:
+    /// - system.errors
+    /// - system.events
+    settings.system_events_show_zero_values = true;
+
+    Words system_tables;
+    fetch(connection, timeouts, "SHOW TABLES IN system", settings, system_tables);
+
+    std::stringstream query; // STYLE_CHECK_ALLOW_STD_STRING_STREAM
+    query << "SELECT DISTINCT arrayJoin(extractAll(name, '[\\\\w_]{2,}')) AS res FROM (";
+
+    std::vector<std::vector<String>> system_column_tables =
+    {
+        {"name", "functions"},
+        {"name", "table_engines"},
+        {"name", "formats"},
+        {"name", "table_functions"},
+        {"name", "data_type_families"},
+        {"name", "merge_tree_settings"},
+        {"name", "settings"},
+        {"cluster", "clusters"},
+        {"name", "errors"},
+        {"event", "events"},
+        {"metric", "asynchronous_metrics"},
+        {"metric", "metrics"},
+        {"macro", "macros"},
+        {"policy_name", "storage_policies"},
+    };
+
+    auto first = true;
+    for (auto & column_table : system_column_tables)
+    {
+        auto it = std::find(system_tables.begin(), system_tables.end(), column_table[1]);
+        if (it == system_tables.end())
+            continue;
+
+        if (!first)
+            query << " UNION ALL ";
+
+        first = false;
+        query << "SELECT " << column_table[0] << " FROM system." << column_table[1];
```

Comment:
FWIW I've already have similar patch - #21943, that check tables for existence, but only if the version of the client and server does not match (and a switch for this)

Reply:
Wow, I did not notice that.  Your patch looks great, should I close this pr?

Reply:
I tend to check the version, but I realize that some system tables may be dropped manually.

Reply:
>I tend to check the version, but I realized that some system tables may be dropped manually.

All tables that is used for completion are virtual tables, and you cannot DROP such tables.
Also one more query for completion is not a problem I guess

Reply:
How about using the same approach as in #21943 for list of available completions, i.e.:
- list of queries
- parse the database.table from it for checking
- use erase_if and boost::join for cleaner code
It will allow not to think about any code around when someone will want to add new bits into suggestion.

Or I can rewrite my patch to use one `SHOW TABLES in system` query.

Reply:
Ok,  I think to rewrite your patch to use  `SHOW TABLES in system` is much better. Because we have thousands tables in prod env.

And I think you should reuse the fetch function as I did in your patch.

Reply:
>And I think you should reuse the fetch function as I did in your patch.

Indeed, it was duplicated because initially it was different, thanks

>Ok, I think to rewrite your patch to use SHOW TABLES in system is much better. Because we have thousands tables in prod env.

You have thousands of tables for suggestion in prod env? :)